### DISCUSSION TITLE
Advice on dynamic choice node

### DISCUSSION THREAD

**Discussion by AckslD** [L3MON4D3/LuaSnip#358] [Q&A] [answered]
I'm looking for advice on a snippet I'm trying to write. I trying to get better at writing luasnip snippets so any advice is appreciated. I'm basically trying to write a small snippet for adding more plugins to my neovim config, which usually has something of the form:
```lua
use { -- LuaSnip {{{
  'L3MON4D3/LuaSnip',
  setup = 'require("plugin_settings.luasnip").setup()',
  config = 'require("plugin_settings.luasnip").config()',
} -- }}}
```
so I started with the following:
```lua
s('u', fmt([[
  use {{ -- {name} {{{{{{
      '{plugin}',
      setup = 'require("plugin_settings.{module}").setup()',
      config = 'require("plugin_settings.{module}").config()',
  }} -- }}}}}}]], {
  plugin = i(1),
  name = d(2, function(plugin)
    plugin = plugin[1][1]
    local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
    return sn(nil, t(name))
  end, {1}),
  module = f(function(name)
    name = name[1][1]
    return name:gsub([[%.lua]], '')
  end, {2}),
}))
```
which works pretty well.

Then I thought I wanted to make the `setup` and `config` keys optional using some `choiceNode`. But not sure what's the best approach to take here. I could maybe do something like:
```lua
s('u', c(1, {
  fmt([[
    use {{ -- {name} {{{{{{
        '{plugin}',
    }} -- }}}}}}]], {
    plugin = i(1),
    name = d(2, function(plugin)
      plugin = plugin[1][1]
      local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
      return sn(nil, t(name))
    end, {1}),
  }),
  fmt([[
    use {{ -- {name} {{{{{{
        '{plugin}',
        setup = 'require("plugin_settings.{module}").setup()',
        config = 'require("plugin_settings.{module}").config()',
    }} -- }}}}}}]], {
    plugin = i(1),
    name = d(2, function(plugin)
      plugin = plugin[1][1]
      local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
      return sn(nil, t(name))
    end, {1}),
    module = f(function(name)
      name = name[1][1]
      return name:gsub([[%.lua]], '')
    end, {2}),
  }),
}))
```
but this has two disadvantages:
* There is a lot of duplication between the two choices, one can of course move some things out to variables but still feels like there should be a better approach.
* Maybe I'm trying to do something which is not possible in `luasnip` but with the above, if I start typing something in the insert node and then change the choice then what I inserted is gone. This is I guess because it's actually not one insert node shared between the choices but instead two different ones. Also change choice a second time seems to not work.

I was thinking it there could be a way to have an internal choice node between nothing and the `config/setup` fields, ie something like:
```lua
s('u', fmt([[
  use {{ -- {name} {{{{{{
      '{plugin}',
  {settings}}} -- }}}}}}]], {
  plugin = i(1),
  name = d(2, function(plugin)
    plugin = plugin[1][1]
    local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
    return sn(nil, t(name))
  end, {1}),
  settings = c(3, {
    t(''),
    fmt([[
      setup = 'require("plugin_settings.{module}").setup()',
      config = 'require("plugin_settings.{module}").config()',
    ]], {
      module = f(function(name)
        name = name[1][1]
        return name:gsub([[%.lua]], '')
      end, {2}),
    }),
  }),
}))
```
But then one needs to jump to the choice node to change it and also seems to only be possible to change once.

Let me know if these types of discussion points don't make sense here. I'm mostly looking for some advice from more experienced snippet writers :)

**Comment by L3MON4D3** [comment, marked as answer]
> if I start typing something in the insert node and then change the choice then what I inserted is gone.

Ouh, here you could use `restoreNode`, it allows you to keep some part of the snippet across changing choice and re-executing a `dynamicNode`.

I can't tell right away why changeChoice won't work the second time in either snippet, that's a bit weird.

Another way: use [this](https://github.com/L3MON4D3/LuaSnip/wiki/Misc#dynamicnode-with-user-input) to generate the entire snippet in a function, then you could use the keybinds shown in the wiki to cycle between different configurations, with minimal duplication.

**Reply by AckslD** [reply]
Thanks a lot for the pointer to `restoreNode`! Worked perfectly :rocket: 

![luasnip_restore](https://user-images.githubusercontent.com/23341710/160884313-c352fd40-ff76-4d07-a9af-6132b958a987.gif)

**Reply by AckslD** [reply]
Regarding the other issue I had, that was just a mistake in my config, I had only the change choice mapped in insert mode and not in select mode :see_no_evil:

**Reply by AckslD** [reply]
For reference, here's the final snippet (which could be written a bit more cleanly)

```lua
            s('u', c(1, {
              fmt([[
                use {{ -- {name} {{{{{{
                    '{plugin}',
                }} -- }}}}}}]], {
                  plugin = r(1, "plugin"),
                  name = d(2, function(plugin)
                    plugin = plugin[1][1]
                    local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
                    return sn(nil, t(name))
                  end, {1}),
                }),
              fmt([[
                use {{ -- {name} {{{{{{
                    '{plugin}',
                    setup = 'require("plugin_settings.{module}").setup()',
                    config = 'require("plugin_settings.{module}").config()',
                }} -- }}}}}}]], {
                  plugin = r(1, "plugin"),
                  name = d(2, function(plugin)
                    plugin = plugin[1][1]
                    local name = plugin:gsub('.*/', ''):gsub('nvim%-', '')
                    return sn(nil, t(name))
                  end, {1}),
                  module = f(function(name)
                    name = name[1][1]
                    return name:gsub([[%.lua]], '')
                  end, {2}),
                }),
            }), {
              stored = {plugin = i(1)}
            })
```

**Reply by L3MON4D3** [reply]
> Thanks a lot for the pointer to `restoreNode`! Worked perfectly 🚀
> 
> ![luasnip_restore](https://user-images.githubusercontent.com/23341710/160884313-c352fd40-ff76-4d07-a9af-6132b958a987.gif)

Not sure if you've checked the doc on choiceNode, there's a setting to also restore the cursor-position, Imo that's really smooth.
(we ought to make that a global setting though)

**Reply by AckslD** [reply]
Oh awesome, yeah that's much better!

**Comment by L3MON4D3** [comment]
> Let me know if these types of discussion points don't make sense here. I'm mostly looking for some advice from more experienced snippet writers :)

Oh, I think posts like this are quite useful, as long as they are also accessible to other users, might serve as inspiration for more complicated snippets :D
(although for most complicated snippets the answer seems to be "use that node from the wiki" xD)
